src/main/java/io/nats/client/impl/NatsMessageConsumer.java

```diff
@@ -53,9 +53,14 @@ class NatsMessageConsumer extends NatsMessageConsumerBase implements PullManager
     public void heartbeatError() {
         try {
             // just close the current sub and make another one.
-            // this could go on endlessly
-            lenientClose();
-            doSub();
+            // this could go on endlessly - unless the user had called stop
+            if (stopped.get()) {
+                finishAndClose();
+            }
+            else {
```

Comment:
If a heartbeat error occurs, we would usually just try again. But if the user had called stop, they've told us that we are finished.